alacritty/src/display/damage.rs

```diff
@@ -0,0 +1,345 @@
+
+        let cursor_point = term.grid().cursor.point;
+        let cursor = TermRenderableCursor { point: cursor_point, shape: Default::default() };
+
+        let term_damage = term.damage();
+        // Ensure it's not damaged.
+        term_damage.take_fully_damaged();
+        let mut renderer_damage = RendererDamage::new(term_damage, size, cursor, None, None, 0);
+
+        let mut cursor_line = DamagedLine::undamaged(num_cols);
+        cursor_line.expand(cursor_point.column.0, cursor_point.column.0);
+        let cursor_rect =
+            RendererDamage::rect_for_line(&size, cursor_point.line.0 as usize, cursor_line);
+        let cursor_rect = renderer_damage.overdamage(cursor_rect);
+
+        assert_eq!(renderer_damage.next(), Some(cursor_rect));
+        assert_eq!(renderer_damage.next(), None);
+    }
+
+    #[test]
+    fn damage_shaping() {
+        let size = SizeInfo::new(10., 100., 1., 1., 0., 0., false);
+        let mut term = Term::new(&UiConfig::default().terminal_config, size, ());
+        let size: SizeInfo<u32> = size.into();
+        let num_cols = term.columns();
+
+        let cursor_point = term.grid().cursor.point;
+        let cursor = TermRenderableCursor { point: cursor_point, shape: Default::default() };
+
+        let term_damage = term.damage();
+        // Ensure it's not damaged.
+        term_damage.take_fully_damaged();
+        let first_rect_line = 32;
+        term_damage.damage_line(first_rect_line - 1, 3, 6);
+        term_damage.damage_line(first_rect_line, 0, 6);
+        term_damage.damage_line(first_rect_line + 1, 4, 6);
+
+        let second_rect_line = (size.height() - 2) as usize;
+        term_damage.damage_line(second_rect_line, 3, 6);
+        term_damage.damage_line(second_rect_line + 1, 4, 9);
+        let mut renderer_damage = RendererDamage::new(term_damage, size, cursor, None, None, 0);
+
+        let mut cursor_line = DamagedLine::undamaged(num_cols);
+        cursor_line.expand(cursor_point.column.0, cursor_point.column.0);
+        let cursor_rect =
+            RendererDamage::rect_for_line(&size, cursor_point.line.0 as usize, cursor_line);
+        let cursor_rect = renderer_damage.overdamage(cursor_rect);
+
+        let mut first_rect = RendererDamage::rect_for_line(
+            &size,
+            first_rect_line,
+            renderer_damage.terminal_damage[first_rect_line],
+        );
+        first_rect.y -= size.cell_height();
+        first_rect.height += 2 * size.cell_height();
+        first_rect = renderer_damage.overdamage(first_rect);
+
+        let mut second_rect = RendererDamage::rect_for_line(
+            &size,
+            second_rect_line,
+            renderer_damage.terminal_damage[second_rect_line],
+        );
+        second_rect.y -= size.cell_height();
+        second_rect.height += size.cell_height();
+        second_rect.width = size.width() - second_rect.x;
+        second_rect = renderer_damage.overdamage(second_rect);
```

Comment:
There's **a lot** of complexity and math in this test. I'm not saying it justifies removal of the test but otoh I'm not able to easily verify this is **correct** either just by looking at it. There could be a billion off-by-ones in there. So if I ever had to make a big change and this test would bother me, I'd probably just toss it.

Just in case you have ideas on how to possibly clean this up, it certainly would be a great candidate.